File: JPP/Lab_4/problem.go
package main

import (
	"fmt"
	"math/rand"
	"sync"
	"time"
)

type Philosopher struct {
	id           int
	leftFork     *sync.Mutex
	rightFork    *sync.Mutex
	mealsToEat   int
	mealsEaten   int
	logLock      *sync.Mutex
}

func (p *Philosopher) think() {
	p.logLock.Lock()
	fmt.Printf("Filozof %d myśli.\n", p.id + 1)
	p.logLock.Unlock()
	time.Sleep(time.Duration(rand.Intn(3)+1) * time.Second)
}

func (p *Philosopher) eat() {
	var firstFork, secondFork *sync.Mutex

	if p.id%2 == 0 {
		firstFork, secondFork = p.leftFork, p.rightFork
	} else {
		firstFork, secondFork = p.rightFork, p.leftFork
	}

	firstFork.Lock()
	p.logLock.Lock()
	fmt.Printf("Filozof %d podniósł pierwszy widelec.\n", p.id + 1)
	p.logLock.Unlock()

	secondFork.Lock()
	p.logLock.Lock()
	fmt.Printf("Filozof %d podniósł drugi widelec.\n", p.id + 1)
	p.logLock.Unlock()

	p.mealsEaten++
	p.logLock.Lock()
	fmt.Printf("Filozof %d je posiłek %d.\n", p.id + 1 , p.mealsEaten)
	p.logLock.Unlock()
	time.Sleep(time.Duration(rand.Intn(2)+1) * time.Second)

	secondFork.Unlock()
	firstFork.Unlock()

	p.logLock.Lock()
	fmt.Printf("Filozof %d odkłada widelce.\n", p.id + 1)
	p.logLock.Unlock()
}

func (p *Philosopher) dine(wg *sync.WaitGroup) {
	defer wg.Done()
	for p.mealsEaten < p.mealsToEat {
		p.think()
		p.eat()
	}
}

func main() {
	numPhilosophers := 6
	mealsToEat := 3

	var forks = make([]*sync.Mutex, numPhilosophers)
	for i := 0; i < numPhilosophers; i++ {
		forks[i] = &sync.Mutex{}
	}

	var logLock sync.Mutex

	var philosophers = make([]*Philosopher, numPhilosophers)
	var wg sync.WaitGroup

	for i := 0; i < numPhilosophers; i++ {
		philosophers[i] = &Philosopher{
			id:         i,
			leftFork:   forks[i],
			rightFork:  forks[(i+1)%numPhilosophers],
			mealsToEat: mealsToEat,
			logLock:    &logLock,
		}
		wg.Add(1)
		go philosophers[i].dine(&wg)
	}

	wg.Wait()
	fmt.Println("Wszyscy filozofowie zakończyli jedzenie.")
}

File: JPP/Lab_4/problem.py
import threading
import time
import random


class Philosopher(threading.Thread):
    def __init__(self, id, left_fork, right_fork, meals_to_eat, log_lock):
        threading.Thread.__init__(self)
        self.id = id
        self.left_fork = left_fork
        self.right_fork = right_fork
        self.meals_to_eat = meals_to_eat
        self.log_lock = log_lock
        self.meals_eaten = 0

    def run(self):
        while self.meals_eaten < self.meals_to_eat:
            self.think()
            self.eat()

    def think(self):
        with self.log_lock:
            print(f"Filozof {self.id + 1} myśli.")
        time.sleep(random.uniform(1, 3))

    def eat(self):
        left, right = self.left_fork, self.right_fork

        if self.id % 2 == 0:
            first, second = left, right
        else:
            first, second = right, left

        with first:
            with self.log_lock:
                print(f"Filozof {self.id + 1} podniósł pierwszy widelec.")
            with second:
                with self.log_lock:
                    print(f"Filozof {self.id +1 } podniósł drugi widelec.")
                self.meals_eaten += 1
                with self.log_lock:
                    print(f"Filozof {self.id + 1} je posiłek {self.meals_eaten}.")
                time.sleep(random.uniform(1, 2))

        with self.log_lock:
            print(f"Filozof {self.id + 1} odkłada widelce.")


def main():
    num_philosophers = 5
    meals_to_eat = 3
    forks = [threading.Lock() for _ in range(num_philosophers)]
    log_lock = threading.Lock()

    philosophers = [
        Philosopher(
            i, forks[i], forks[(i + 1) % num_philosophers], meals_to_eat, log_lock
        )
        for i in range(num_philosophers)
    ]

    for philosopher in philosophers:
        philosopher.start()

    for philosopher in philosophers:
        philosopher.join()

    print("Wszyscy filozofowie zakończyli jedzenie.")


if __name__ == "__main__":
    main()
